master/internal/stream/experiments.go

```diff
@@ -0,0 +1,278 @@
+		})
+		return out, nil
+	}
+	// step 0: get user's permitted access scopes
+	accessMap, err := AuthZProvider.Get().GetExperimentStreamableScopes(ctx, user)
+	if err != nil {
+		return nil, err
+	}
+	var accessScopes []model.AccessScopeID
+	for id, isPermitted := range accessMap {
+		if isPermitted {
+			accessScopes = append(accessScopes, id)
+		}
+	}
+
+	// step 1: calculate all ids matching this subscription
+	q := db.Bun().NewSelect().
+		Table("experiments").
+		Column("id").
+		Join("JOIN projects p ON experiments.project_id = p.id").
+		OrderExpr("experiments.id ASC")
+
+	q = permFilter(q, accessMap, accessScopes)
+
+	// Ignore tmf.Since, because we want appearances, which might not be have seq > spec.Since.
+	ws := stream.WhereSince{Since: 0}
+	if len(spec.ExperimentIds) > 0 {
+		ws.Include("experiment_id in (?)", bun.In(spec.ExperimentIds))
+	}
+	q = ws.Apply(q)
+
+	var exist []int64
+	err = q.Scan(ctx, &exist)
+	if err != nil && errors.Cause(err) != sql.ErrNoRows {
+		log.Errorf("error: %v\n", err)
+		return nil, err
+	}
+
+	// step 2: figure out what was missing and what has appeared
+	missing, appeared, err := stream.ProcessKnown(known, exist)
+	if err != nil {
+		return nil, err
+	}
+
+	// step 3: hydrate appeared IDs into full TrialMsgs
+	var expMsgs []*ExperimentMsg
+	if len(appeared) > 0 {
+		query := getExperimentMsgsWithWorkspaceID(expMsgs).
+			Where("exp_msg.id in (?)", bun.In(appeared))
+		query = permFilter(query, accessMap, accessScopes)
+		err := query.Scan(ctx, &expMsgs)
+		if err != nil && errors.Cause(err) != sql.ErrNoRows {
+			log.Errorf("error: %v\n", err)
+			return nil, err
+		}
+	}
+
+	// step 4: emit deletions and updates to the client
+	out = append(out, stream.DeleteMsg{
+		Key:     ExperimentsDeleteKey,
+		Deleted: missing,
+	})
+	for _, msg := range expMsgs {
+		out = append(out, stream.UpsertMsg{JSONKey: ExperimentsUpsertKey, Msg: msg})
+	}
+	return out, nil
+}
+
+// ExperimentCollectSubscriptionModMsgs scrapes the database when a
+// user submits a new TrialSubscriptionSpec for initial matches.
+func ExperimentCollectSubscriptionModMsgs(ctx context.Context, addSpec ExperimentSubscriptionSpec) (
+	[]interface{}, error,
+) {
+	if len(addSpec.ExperimentIds) == 0 {
+		return nil, nil
+	}
+	var expMsgs []*ExperimentMsg
+	q := getExperimentMsgsWithWorkspaceID(expMsgs)
+
+	// Use WhereSince to build a complex WHERE clause.
+	ws := stream.WhereSince{Since: addSpec.Since}
+	if len(addSpec.ExperimentIds) > 0 {
+		ws.Include("experiment_id in (?)", bun.In(addSpec.ExperimentIds))
+	}
+	q = ws.Apply(q)
+
+	err := q.Scan(ctx)
+	if err != nil && errors.Cause(err) != sql.ErrNoRows {
+		log.Errorf("error: %v\n", err)
+		return nil, err
+	}
+
+	var out []interface{}
+	for _, msg := range expMsgs {
+		out = append(out, msg.UpsertMsg())
+	}
+	return out, nil
+}
+
+// ExperimentFilterMaker tracks the trial and experiment id's that are to be filtered for.
+type ExperimentFilterMaker struct {
+	ExperimentIds map[int]bool
```

Comment:
nit (non-blocking): it might be worth replacing these maps with: 
```suggestion
	ExperimentIds map[int]struct{}
```

This way we can conserve some space; however, I don't know how beneficial this would really be and if it's worth making the code more difficult to read in order to do so. 

So more food for thought, rather than an action item